File: day-8/go/part1.go
package day8

import (
	"fmt"
	"os"
	"strings"
	"time"
)

type Day8 struct{}

type Direction struct {
	Left  string
	Right string
}

func (d Day8) Part1(filename *string) string {
	// start clock
	start := time.Now()

	content, _ := os.ReadFile(*filename)
	lines := strings.Split(string(content), "\n")

	var network = map[string]Direction{}
	var instructions string

	// create network in form of map
	for idx, line := range lines {
		if idx == 1 {
			continue
		}

		if idx == 0 {
			instructions = line
			continue
		}

		var parts = strings.Split(line, " = ")
		var direction = strings.Split(parts[1], ", ")
		var left_direction = direction[0][1:]
		var right_direction = direction[1][:3]
		network[parts[0]] = Direction{left_direction, right_direction}
	}

	// Start at the first element of the network
	var steps = 0
	var curr_element = "AAA"
	var instruction_idx = 0
	for curr_element != "ZZZ" {
		var directions = network[curr_element]
		var curr_direction = instructions[instruction_idx]
		if string(curr_direction) == "L" {
			curr_element = directions.Left
		} else {
			curr_element = directions.Right
		}

		if instruction_idx == len(instructions)-1 {
			instruction_idx = 0
		} else {
			instruction_idx++
		}

		steps++
	}

	fmt.Println(time.Since(start))
	return fmt.Sprint(steps)
}

File: day-8/cpp/part1.cpp
// To run do: c++ part1.cpp -Wall -std=c++17 -o output1.out && ./output1.out
#include <string>
#include <vector>
#include <fstream>
#include <iostream>
#include <map>
#include "../../utils/utils.cpp"

using namespace std;

int main()
{
    // Initialize a vector to store the file input and set the input file name.
    std::string fileName = "../input.txt";
    std::vector<std::string> lines = read_file(fileName);

    map<string, tuple<string, string>> network;
    string instructions;

    for (int i = 0; i < lines.size(); i++)
    {
        if (i == 1)
        {
            continue;
        }

        if (i == 0)
        {
            instructions = lines[i];
            continue;
        }

        vector<string> parts = splitter(lines[i], " = ");
        vector<string> direction = splitter(parts[1], ", ");
        string left_direction = splitter(direction[0], "(")[1];
        string right_direction = splitter(direction[1], ")")[0];
        network[parts[0]] = tuple(left_direction, right_direction);
    }

    int steps = 0;
    string curr_element = "AAA";
    int instruction_idx = 0;

    while (curr_element != "ZZZ")
    {
        tuple<string, string> directions = network[curr_element];
        char instruction = instructions[instruction_idx];

        if (instruction == 'L')
        {
            curr_element = get<0>(directions);
        }
        else
        {
            curr_element = get<1>(directions);
        }

        if (instruction_idx == instructions.length() - 1)
        {
            instruction_idx = 0;
        }
        else
        {
            instruction_idx += 1;
        }

        steps += 1;
    }

    // Print the result.
    std::cout << "Result: " << steps << std::endl;
}
